### PR TITLE
wip implementation of multiple connections

### LINKED ISSUE #64: Store audit data in a separate database
Great extension, thanks!

I'd like to store audit data in a separate database, potentially on a different physical server. 

Is such a thing even possible ? 

If it's just supplying the config option for new entity manager and using it in LogRevisionsListener, then I can do it and create a pull request, it just seems too easy to not be in there already :)

### PR DESCRIPTION
**Pull request by andrewtch**
Closes #64 
Closes #118

### CODE REVIEW

File: src/SimpleThings/EntityAudit/EventListener/LogRevisionsListener.php [hunk 1]
```diff
@@ -225,7 +225,7 @@ public function postUpdate(LifecycleEventArgs $eventArgs)
     public function onFlush(OnFlushEventArgs $eventArgs)
     {
         $this->em = $eventArgs->getEntityManager();
-        $this->conn = $this->em->getConnection();
+        $this->conn = $this->config->getConnection() ? $this->config->getConnection() : $this->em->getConnection();
```
Comment (DavidBadura) [inline src/SimpleThings/EntityAudit/EventListener/LogRevisionsListener.php]: the shortcut for

```php
$this->conn = $this->config->getConnection() ? $this->config->getConnection() : $this->em->getConnection();
```

is

```php
$this->conn = $this->config->getConnection() ?: $this->em->getConnection();
```

I think it's also easier to read :wink:

**Comment by DavidBadura** [comment]
:+1: i like it

File: src/SimpleThings/EntityAudit/AuditReader.php [hunk 2]
```diff
@@ -149,7 +156,8 @@ public function __construct(EntityManager $em, AuditConfiguration $config, Metad
         $this->em = $em;
         $this->config = $config;
         $this->metadataFactory = $factory;
-        $this->platform = $this->em->getConnection()->getDatabasePlatform();
+        $this->connection = $config->getConnection() ? $config->getConnection() : $this->em->getConnection();
```
Comment (aivus) [inline src/SimpleThings/EntityAudit/AuditReader.php]: `$config->getConnection() ?: $this->em->getConnection();`

**Comment by andrewtch** [comment]
OK, Im always afraid of short terniary somehow)  some more tests, especially with collections are needed though.

**Comment by aivus** [comment]
:+1: thanks for this implementation

**Comment by aivus** [comment]
Don't forget to change `auditable.xml` and DI configuration, please. Thanks

**Comment by andrewtch** [comment]
That's why it's still WIP. I'm porting tests currently.

**Comment by andrewtch** [comment]
Ouch, hopefully finished.

@aivus , can you please test it in your environment?
@DavidBadura , can you check the code, especially at https://github.com/simplethings/EntityAudit/pull/120/files#diff-526bd1905ced8959412a89500653b48aR62 - is it a correct approach or a sh**code? Because else we'd need to use container expressions (like @doctrine.getConnection(audit)) or introduce whole Repository, which would make problems for no-symfony projects.

**Comment by DavidBadura** [comment]
tonight i will send a pull request :wink:

**Comment by andrewtch** [comment]
ok, i'll wait then (though you can just offer a better approach if you can explain it)

**Comment by andrewtch** [comment]
@aivus , can you pelase test this?

**Comment by andrewtch** [comment]
@aivus , ping!

**Comment by aivus** [comment]
I will test it on this week

File: src/SimpleThings/EntityAudit/DependencyInjection/SimpleThingsEntityAuditExtension.php [hunk 3]
```diff
@@ -58,5 +56,13 @@ public function load(array $configs, ContainerBuilder $container)
         if (true === $config['listener']['current_username']) {
             $loader->load('current_username.xml');
         }
+
+        if ($config['audit_connection']) {
+            $container->getDefinition('implethings_entityaudit.config')
```
Comment (aivus) [inline src/SimpleThings/EntityAudit/DependencyInjection/SimpleThingsEntityAuditExtension.php]: It should be "simplethings_entityaudit.config"

**Comment by aivus** [comment]
I have `revision` connection, but when I try set
`audit_connection: revision` in `simple_things_entity_audit` section in config.yml i got:
`The service "simplethings_entityaudit.config" has a dependency on a non-existent service "revision".`

How can I set connection for `audit_connection`?

File: src/SimpleThings/EntityAudit/DependencyInjection/SimpleThingsEntityAuditExtension.php [hunk 4]
```diff
@@ -58,5 +56,13 @@ public function load(array $configs, ContainerBuilder $container)
...
+                ->addMethodCall('setAuditConnection', array(
+                        new Reference($config['audit_connection'])
```
Comment (aivus) [inline src/SimpleThings/EntityAudit/DependencyInjection/SimpleThingsEntityAuditExtension.php]: Hm, it works when I use:
`new Reference('doctrine.dbal.' . $config['audit_connection'] . '_connection')`

**Comment by aivus** [comment]
When I use `app/console doctrine:schema:update --dump-sql` I expect get sql, but get created audit tables in database.

**Comment by andrewtch** [comment]
Yes, it's not a bug, it was intended to use doctrine.dbal.CONNECTIONNAMEHERE_connection as a parameter.

However, there is indeed a big hole in schema generation. @DavidBadura , can you take a look on issue below?

The thing is that we need to create audit tables for a different connections; as you can see in ```CreateSchemaListener``` we need to determine which connection we're using.

For ```postGenerateSchema``` it's a simple as ```$entArgs->em->getConnection()``` - this issue is that the connection is not accessible in ```postGenerateSchemaTable```. Do you know a way to get a connection instance in ```postGenerateSchemaTable``` so we can do something like:

```php
if ($this->config->getAuditConnection() && $em->getConnection() != $this->config->getAuditConnection()) {
    //skip tables for uknown connections

    return;
}
```

?

**Comment by andrewtch** [comment]
I've made a very basic implementation of multiple connections that cleans up uneeded tables for connections that are not audit; however, we still need to add support of multiple connections on service definition level.

@DanielBadura , can you review?

**Comment by DavidBadura** [comment]
I have no time this week. I will review at the weekend. :-/

**Comment by andrewtch** [comment]
David? )

**Comment by andrewtch** [comment]
@DavidBadura , ping?

**Comment by DavidBadura** [comment]
Hey @andrewtch it looks very good. What do you mean with "we still need to add support of multiple connections on service definition level."?

**Comment by andrewtch** [comment]
Em, you see...

In postGenerateSchemaTable I create audit tables. These are created for the main connection, not for the audited. 

Yes, there's a workaround in postGenerateSchema that skips all "normal" tables from creation, but this is way ugly since we force the entities to be bound to audit connection (thus having 2x EntityManagers).

 So the question is how to correctly schedule table creation for one connection when actually creating entities for another?

**Comment by Filoz** [comment]
Great job @andrewtch !
will it be merged?

**Comment by andrewtch** [comment]
Closing as obsolete

**Comment by bendavies** [comment]
@andrewtch was the only problem with this that the SchemaListener doesn't work?

**Comment by andrewtch** [comment]
@bendavies , yes, but it whould require majot workaround to deal with this.

**Comment by bendavies** [comment]
@andrewtch yeah, that's what i thought. did you have any ideas? i couldn't see anything quickly...

**Comment by andrewtch** [comment]
I also can't except for some reflection woodoo and messing up schemalistener (see https://github.com/simplethings/EntityAudit/pull/120#issuecomment-95834844 and https://github.com/simplethings/EntityAudit/pull/120#issuecomment-83614356)

**Comment by bendavies** [comment]
Looks like to me we can't use a schema listener if it's on it's own new connection.
I would say we need to use a command to list the sql needed to create the diff much like how `init:acl` works in the symfony security bundle.

**Pull request closed without merging**